pkg/plugin/plugin.go

```diff
@@ -359,8 +361,44 @@ func CmdAdd(args *skel.CmdArgs) error {
 			ipc.Interface = current.Int(0)
 		}
 
+		// wait until OF port link state becomes up. This is needed to make
+		// gratuitous arp for args.IfName to be sent over ovs bridge
+		for i := 0; i < 5; i++ {
+			portState, err := ovsDriver.GetOFPortOpState(hostIface.Name)
+			if err != nil {
+				log.Printf("error in retrieving port %s state: %v", hostIface.Name, err)
+			} else {
+				if portState == "up" {
+					break
+				}
+			}
+			if i == 4 {
+				log.Printf("The OF port %s state is not up until this time", hostIface.Name)
+			} else {
+				time.Sleep(500 * time.Millisecond)
+			}
+		}
```

Comment:
I don't feel comfortable with this code in. Issue is that it never fails, it only logs an error. And we don't have a test covering it. Due to that, this can just stop working and we will never notice.

I imagine testing this may become quite difficult. For me, it would be enough if we extended the existing IPAM test with an assertion making sure that the output never logs these issues (although it would be quite easy to get a false positive then).

Any other idea how to test it?

Or could we make the new code less forgiving and hard fail in case things don't go as expected?

Reply:
I see no problem in letting the function fail hard in case the link state of the veth or VF representor port doesn't come up. But I wonder what is the approach in other forwarders in that case.

Reply:
@JanScheurich the code is now changed to use fail fast approach when link state is not up until given retries/interval.  This check may not be needed in CNIs like [macvlan ](https://github.com/containernetworking/plugins/blob/master/plugins/main/macvlan/macvlan.go#L318)because that doesn't have something OF port which comes operationally up bit later after veth link is up.